# _Ursa miner_

![Ursa minor by Hevelius](https://upload.wikimedia.org/wikipedia/commons/2/2a/Ursa_Minor_Hevelius.jpg)

_Ursa miner_ is a small tool to enable the use of bootstrap data generated by the cutting-edge RNA-sequencing tool [kallisto](http://pachterlab.github.io/kallisto) from Lior Pachter's group.

## Installation

	git clone https://github.com/blachlylab/ursaminer.git
	chmod +x ursaminer/ursa.py
	export PATH=$PATH:$(pwd)/ursaminer


### Requirements

_Ursa miner_ requires python 2.7 or greater and the `numpy`, `pandas`, and `h5py` packages. If you are using the Anaconda distribution of python you can install with your package manager. Otherwise, we find python virtual environments to be the best way to install dependencies:
        
    virtualenv ursaminer
    source ursaminer/bin/activate
    pip install numpy
    pip install pandas
    pip install h5py

## Usage

`ursa.py {describe | extract} /path/to/abundance.h5`

### Describe

The `describe` function emits the same data present in kallisto's `run_info.json` file, but sources it from the HDF5 instead. This ensures the HDF5 store is independent / you only need keep up with one file per run.

    $ ursa.py describe abundance.h5
    
    Kallisto version     : 0.42.2.1
    Index version        : 10
    Number of targets    : 121375
    Number of bootstraps : 100
    Kallisto started     : Thu Aug 13 19:22:09 2015
    Command              : kallisto quant -i /home/you/indices/kallisto_index -o OUTDIR --bias -t 32 -b 100 fastqgz/XJB05_R1.fastq.gz fastqgz/XJB05_R2.fastq.gz

### Extract

The `extract` function examines the HDF5 store for the presence or absence of bootstrap resampled estimated counts. Finding none, it outputs a file `ursa.csv` that is identical to kallisto's `abundance.tsv`. If resamplings are detected, _Ursa miner_ compiles all together and computes summary statistics on the estimated counts:

* mean
* std dev
* std err of mean
* minimum
* median
* maximum

_Ursa miner_ adds computes and adds a final column, `tpm` (transcripts per million), and writes `ursa.csv` to the current working directory.

In this way, the HDF5 file can stand alone; it contains all the information in both the JSON file that describes run information as well as all the information in `abundance.tsv`. With _Ursa miner_ there is no need to keep track of more than one secondary data file per sample. 

    $ ursa.py extract abundance.h5
    
    Loading data...
    Computing statistics...
    Writing ursa.csv...


## FAQ

### You misspelled _Ursa minor_

Nope. That's also a statement, not a question.
If the pun is not that obvious I am going to have to re-think the name...

### Reading CSV files stinks. TSV is way better.

Also not a question. But I suggest:

`column -ts, ursa.csv |less -S` 

for convenient table viewing.

### Why didn't you use `pandas` built-in HDF5 support?

The HDF5 support in `pandas` is build on `pytables`; `pytables` unfortunately does not support the HDF5 format used by kallisto. The excellent [h5py](https://github.com/h5py/h5py) library, however, does.

### How is the TPM calculated?

* For experiments in which no bootstrapping was done, directly from the only available est\_counts (naturally).
* For experiments in which bootstrapping was performed, from the mean of est\_counts across all resamplings.

### Can I have TPM calculated from the median / some other quantile of counts?

Not yet, but it would be trivial if you are so desirous. Let us know.

### Why are there not measures of uncertainty for the TPM value?

I need to double check with my statistican, or the Pachter group can chime in, on the correct way to express the confidence bounds surrounding TPM, which is a value derived from the estimated counts that itself has uncertainty measures.

